mlflow/utils/autologging_utils.py

```diff
@@ -260,3 +292,448 @@ def batch_metrics_logger(run_id):
+                "Encountered unexpected error during %s autologging: %s", autologging_integration, e
+            )
+
+        if original_has_been_called:
+            return original_result
+        else:
+            return original(*args, **kwargs)
+
+    wrap_patch(destination, function_name, safe_patch_function)
+
+
+def _validate_args(
+    user_call_args, user_call_kwargs, autologging_call_args, autologging_call_kwargs
+):
+    """
+    Used for testing purposes to verify that, when a patched ML function calls its underlying
+    / original ML function, the following properties are satisfied:
+
+        - All arguments supplied to the patched ML function are forwarded to the
+          original ML function
+        - Any additional arguments supplied to the original function are exception safe (i.e.
+          they are either functions decorated with the `@exception_safe_function` decorator
+          or classes / instances of classes with type `ExceptionSafeClass`
+    """
+
+    def _validate_new_input(inp):
+        """
+        Validates a new input (arg or kwarg) introduced to the underlying / original ML function
+        call during the execution of a patched ML function. The new input is valid if:
+
+            - The new input is a function that has been decorated with `exception_safe_function`
+            - OR the new input is a class with the `ExceptionSafeClass` metaclass
+            - OR the new input is a list and each of its elements is valid according to the
+              these criteria
+        """
+        if type(inp) == list:
+            for item in inp:
+                _validate_new_input(item)
+        elif callable(inp):
+            assert getattr(inp, _ATTRIBUTE_EXCEPTION_SAFE, False), (
+                "New function argument '{}' passed to original function is not exception-safe."
+                " Please decorate the function with `exception_safe_function`.".format(inp)
+            )
+        else:
+            assert hasattr(inp, "__class__") and type(inp.__class__) == ExceptionSafeClass, (
+                "Invalid new input '{}'. New args / kwargs introduced to `original` function"
+                " calls by patched code must either be functions decorated with"
+                "`exception_safe_function`, instances of classes with the `ExceptionSafeClass`"
+                " metaclass safe or lists of such exception safe functions / classes.".format(inp)
+            )
+
+    def _validate(autologging_call_input, user_call_input=None):
+        if user_call_input is None and autologging_call_input is not None:
+            _validate_new_input(autologging_call_input)
+            return
+
+        assert type(autologging_call_input) == type(
+            user_call_input
+        ), "Type of input to original function '{}' does not match expected type '{}'".format(
+            type(autologging_call_input), type(user_call_input)
+        )
+
+        if type(autologging_call_input) == list:
```

Comment:
```suggestion
        if type(autologging_call_input) == tuple:
```

Should we use `tuple` here?

```python
def foo(*args):
    print(type(args))

foo("a", "b")  # -> <class 'tuple'>
```

Reply:
Yes! this was an oversight on my part. I've added tuple support and updated the associated tests to test against tuple input types! Thanks so much!